### DISCUSSION TITLE
무중단 배포 환경 구성

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [woowacourse-teams/2023-pium#358] [개발 논의]
# 서론

cc. @rawfishthelgh @Kim0914 @yeonkkk 

안녕하세요 친애하는 피움 백엔드 팀원 여러분
현재 백엔드 배포 과정에서 서버가 down되고 새롭게 start 되고 있는 상황입니다.

사용자 유치를 약 100명 넘게 한 현 시점에서 사용자 경험에 불편함을 주지 않기 위해 배포 시 접속이 원활하게 유지될 수 있도록 무중단 배포환경을 구성하면 좋겠다는 생각을 하게되었습니다.

# 고가용성

다음 자료는 페이지 로드 속도가 늦어짐에 따라 사용자의 이탈률이 증가한다는 지표입니다.

![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/58b2ac61-f3bd-473d-8e15-11594740cc61)

[출처 - Think with Google](https://www.thinkwithgoogle.com/marketing-strategies/app-and-mobile/mobile-page-speed-new-industry-benchmarks/)

애플리케이션이 아닌 배포 과정이 다운타임에 대한 영향을 미치는 것이 올바르지 못한 상황이라고 생각됩니다.

<img width="857" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/b193c6a9-e5d5-42f6-a207-af41aecd83d2">

[위키백과 - 고가용성 백분율 계산표](https://ko.wikipedia.org/wiki/%EA%B3%A0%EA%B0%80%EC%9A%A9%EC%84%B1#%EB%B0%B1%EB%B6%84%EC%9C%A8_%EA%B3%84%EC%82%B0)

# 서버 사양 확인

현재 WAS 한대를 구동하고 있는 상황에서 사용중인 메모리 양은 다음과 같습니다.

<img width="1035" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/d7f72fe9-6f4f-431e-b252-2328f77322d3">

(스왑메모리 제외) 총 1.8 GB 중 742MB를 상시 사용중이고 이중 Application이 534MB만큼을 사용중입니다.
Application의 접속량에 따른 메모리 추가 사용을 약 10% 정도로 고려해봤을 때 여유를 잡고 약 600MB 정도의 메모리를 사용한다고 생각해볼 수 있을 것 같습니다 :)

이에 새로운 Application을 서버에 한 대 더 띄울경우 약 600MB를 더 사용하게되어 1.4GB 메모리를 상시 사용할 것으로 예상됩니다.

# 무중단 배포 방식

- 롤링 배포: 서버를 여러개로 구성하고 구버전에서 신버전으로 점진적으로 배포하는 방식입니다.
- 블루 / 그린 배포: 다른 서버에 새로운 버전을 띄우고 트래픽을 한번에 변경하는 방식입니다.
- 카나리 배포: 특정 유저에게만 점진적으로 배포하는 방식입니다. 주로 10분에 20% 정도로 트래픽을 분산한다고 합니다.

# 결론

위와 같은 이유들로 무중단 배포를 도입하고 싶다! 라는 의견입니다.
이에 대한 여러분들의 의견이 궁금합니다.

만약 도입을 한다면 서버에 8080 / 8081 두개의 포트로 Blue / Green 배포하는 방식을 생각하고있습니다.

**Comment by Kim0914** [comment]
롤링 배포와 카나리 배포 방식은 구버전/신버전을 함께 배포해야하는 상황이 생길 것 같아 리소스가 많이 들 것 같아요 ㅎㅎ

말씀해주신대로 아직까지 여유 메모리가 있으니, WAS 2개를 포트로 구분하거나 도커로 감싸는 여러 방법을 이용하는 `블루/그린` 방식을 이용하는게 좋다는 생각이 드네요 🫠

**Comment by Choi-JJunho** [comment]
적용 완료로 close하겠습니다.

- https://pium-official.github.io/dont-stop-deploy/
